### PR TITLE
WIP: Convolutional neural network models

### PR DESCRIPTION
**Pull request by arokem**

### CODE REVIEW

File: afqinsight/cnn.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+
+		X = np.swapaxes(X.reshape((subjects, self.channels, self.nodes)), 1, 2)
+
+		if y is not None:
+			return X, y
+		else:
+			return X
+
+	def fit(self, X, y, X_test, y_test):
```
Comment (richford) [inline afqinsight/cnn.py:234]: The inclusion of `X_test` and `y_test` here isn't quite in line with the sklearn estimator API. You could specify the percentage of `X` that you want to use for the test set and pass that through to `ModelBuilder`. Or you could call `train_test_split` here. And for reproducibility, you could add a `random_state` parameter to the class `__init__` and pass that into `train_test_split`.

File: afqinsight/cnn.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+		"""
+		Takes an input dataset X and a series of targets y and trains the CNN.
+		"""
+		# nodes * channels must = X.shape[1]
+		if self.nodes * self.channels != X.shape[1]:
+			raise ValueError("The product nodes and channels is not the correct shape.")
+		# error
+		else:
+			X, y = self._preprocess(X, y)
```
Comment (richford) [inline afqinsight/cnn.py:243]: I think this cold be moved out of the `else` clause. The `if` block will raise an error, so anything that happens after that can proceed as normal.

File: afqinsight/cnn.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+	def predict(self, X):
+		"""
+		Takes an input dataset X and returns the predictions for each example in X.
+		"""
+		X = self._preprocess(X)
+		check_is_fitted(self, 'is_fitted_')
+		pred = self.model_.predict(X).squeeze()
+		return pred
+
+	def score(self, y_test, y_hat):
```
Comment (richford) [inline afqinsight/cnn.py:267]: If you inherit from `RegressorMixin` or `ClassifierMixin` then it will already have a `score` method.

File: afqinsight/cnn.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+		nan_mask = np.logical_not(np.isnan(y_test))
```
Comment (richford) [inline afqinsight/cnn.py:268]: Not sure I understand the requirement for the `NaN` check here. Could you help me understand?

File: afqinsight/tests/test_cnn.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+# Bring your packages onto the path
+
+# Now do your import
+import afqinsight as afqi
+from afqinsight.cnn import CNN
+from afqinsight.datasets import load_afq_data
+
+data_path = op.join(afqi.__path__[0], "data")
+test_data_path = op.join(data_path, "test_data")
+
```
Comment (richford) [inline afqinsight/tests/test_cnn.py:12]: Once the sklearn compatibility is implemented, we should add `check_estimator(CNN())` to the tests. See: https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator

**Review by richford — CHANGES_REQUESTED** [review]
Great work! I made some comments line-by-line. I'd recommend making a `BaseCNN` class that inherits from `BaseEstimator` and then defining two subclasses `CNNClassifier` and `CNNRegressor` that inherit from `ClassifierMixin` and `RegressorMixin`, respectively.

**Comment by richford** [comment]
There are a lot of files in this PR like `untitled_project/oracle.json` and `untitled_project/trial_5223582d2b335dad9cfb5a07c3126556/checkpoints/epoch_0/checkpoint.data-00000-of-00001`. It seems like these are checkpoint files from keras tuner, maybe. Are they supposed to be here or are they the byproduct of a `git add -A`?

**Comment by richford** [comment]
Again, great work overall! I also think this could use some docstring improvements and maybe some more sphinx documentation and examples.

**Comment by richford** [comment]
Some flake8 errors here caught by the CI. I think you might be able to fix all of these with
`black .` run from the root of the git repo.

File: afqinsight/cnn.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+def build_model(hp, conv_layers, input_shape):
+	"""
+	Uses keras tuner to build model - can control # layers, # filters in each layer, kernel size,
+	regularization etc
+
+	"""
+	model = Sequential()
+
+	dense_filters = hp.Int('dense_filters', min_value=32, max_value=512, step=32)
+	model.add(Dense(dense_filters, activation='relu', input_shape=input_shape))
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:23]: Done

File: afqinsight/cnn.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+	This class implements the common sklearn model interface (has a fit and predict function).
+	"""
+
+	def __init__(self, nodes, channels, max_epochs=50, batch_size=32, tuner=None, layers=1, val_split=0.2, **tuner_kwargs):
+		"""
+		Constructs a CNN that uses the given number of nodes, each with a
+		max depth of max_depth.
+		"""
+		# checking nodes is passed as int
+		if not type(nodes) is int:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:166]: Done

File: afqinsight/cnn.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter nodes must be an integer.")
+		else:
+			self.nodes = nodes
+
+		# checking channels is passed as int
+		if not type(channels) is int:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:172]: Done

File: afqinsight/cnn.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter channels must be an integer.")
+		else:
+			self.channels = channels
+
+		# checking layers is passed as int
+		if not type(layers) is int:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:178]: Done

File: afqinsight/cnn.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter layers must be an integer.")
+		else:
+			self.layers = layers
+
+		# checking max epochs is passed as int
+		if not type(max_epochs) is int:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:184]: Done

File: afqinsight/cnn.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter max_epochs must be an integer.")
+		else:
+			self.max_epochs = max_epochs
+
+		if not type(batch_size) is int:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:189]: Done

File: afqinsight/cnn.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter max_epochs must be an integer.")
+		else:
+			self.batch_size = batch_size
+
+		# checking tiner is passed as str or None
+		if type(tuner) is not str and tuner is not None:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:195]: Done

File: afqinsight/cnn.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+			raise TypeError("Parameter tuner must be str.")
+		else:
+			self.tuner = tuner  # tuner can be None (no tuning) BayesianOptimization, Hyperband, or RandomSearch
+
+		# checking val split is passed as float
+		if not type(val_split) is float:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:201]: Done

**Comment by richford** [comment]
@cbarnes7, you mentioned that you addressed some of my comments but I don't see any new commits. Did you push those to the [cbarnes7:cnn](https://github.com/cbarnes7/AFQ-Insight/tree/cnn) branch?

File: afqinsight/cnn.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
+import numpy as np
+from sklearn.utils.validation import check_X_y, check_is_fitted
+from sklearn.metrics import r2_score
+from tensorflow.keras.models import Sequential
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:4]: done

File: afqinsight/cnn.py (hunk 1 above)
Comment (cbarnes7) [reply afqinsight/cnn.py:234]: done

File: afqinsight/cnn.py (hunk 2 above)
Comment (cbarnes7) [reply afqinsight/cnn.py:243]: done

**Comment by cbarnes7** [comment]
> @cbarnes7, you mentioned that you addressed some of my comments but I don't see any new commits. Did you push those to the [cbarnes7:cnn](https://github.com/cbarnes7/AFQ-Insight/tree/cnn) branch?

@richford I just pushed a few of the smaller changes. still working through them

**Comment by arokem** [comment]
@cbarnes7 : I think that you still need to run something like `git push origin cnn` to get the newer commits into the PR.

**Comment by cbarnes7** [comment]
@arokem @richford I did push using `git push origin cnn` but I have bad service (on a roadtrip) so it may not have gone through… I’ll let you know when I’m able to push again

File: afqinsight/cnn.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+		best_model = self._get_best_weights(model, X, y)
+		return best_model
+
+
+class CNN:
+	"""
+	This class implements the common sklearn model interface (has a fit and predict function).
+	"""
+
+	def __init__(self, nodes, channels, max_epochs=50, batch_size=32, tuner=None, layers=1, val_split=0.2, **tuner_kwargs):
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py]: @richford since we aren't making this fully sklearn-compatible, should i adjust __init__ so it doesn't contain logic? or is it fine as is for now?

File: afqinsight/cnn.py [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+		best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]
+
+		# make CNN model using best hyperparameters
+		model = tuner.hypermodel.build(best_hps)
+
+		best_model = self._get_best_weights(model, X, y)
+		return best_model
+
+
+class CNN:
```
[the remark this answers is not in the document]
Comment (richford) [reply afqinsight/cnn.py:266]: Nevermind. We decided to either ignore or postpone an sklearn-compatible API.

File: afqinsight/cnn.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+		# checking val split is passed as float
+		if not type(val_split) is float:
+			raise TypeError("Parameter val_split must be a float.")
+		else:
+			self.val_split = val_split
+		self.tuner_kwargs = tuner_kwargs
+		self.model_ = None
+		self.best_hps_ = None
+
+	def _preprocess(self, X, y=None):
```
[the remark this answers is not in the document]
Comment (richford) [reply afqinsight/cnn.py:327]: Nevermind, see above comment about postponing or ignoring the sklearn-compatible API.

File: afqinsight/cnn.py (hunk 3 above)
Comment (richford) [reply afqinsight/cnn.py:444]: Ignore. We're postponing or ignoring the sklearn-compatible API.

File: afqinsight/tests/test_cnn.py (hunk 5 above)
Comment (richford) [reply afqinsight/tests/test_cnn.py:12]: Ignore this

File: afqinsight/cnn.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (richford) [reply afqinsight/cnn.py]: It's fine as is now.

File: afqinsight/cnn.py [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+
+		y : array-like of shape (n_samples,) or (n_samples, n_targets)
+			Target values
+
+		X_test : array-like of shape (n_samples, n_features)
+			The feature test samples
+
+		y_test : array-like of shape (n_samples,) or (n_samples, n_targets)
+			Target test values
+
```
Comment (richford) [inline afqinsight/cnn.py:395]:
```suggestion
```

Suggest getting rid of these docstrings lines now that you've gotten rid of the input parameters.

File: afqinsight/cnn.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (richford) [reply afqinsight/cnn.py]: @cbarnes7, thanks! However, I don't see `_check_keras()` being called anywhere by any of the other classes.

File: afqinsight/cnn.py [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+	def _preprocess(self, X, y=None):
+		if len(X.shape) > 2:
+			raise ValueError("Expected X to be a 2D matrix.")
+		if y is not None:
+			nan_mask = np.logical_not(np.isnan(y))
+			X = X[nan_mask, :]
+			y = y[nan_mask]
+
+		imp = SimpleImputer(strategy='median')
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:358]: done

File: afqinsight/cnn.py [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+		imp.fit(X)
+		X = imp.transform(X)
+
+		if y is not None:
+			X, y = check_X_y(X, y)
+
+		subjects = X.shape[0]
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:365]: @richford n_subjects is declared by `n_subjects = X.shape[0]` so that would always be correct, but where is n_features coming from?

File: afqinsight/cnn.py [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+			return X, y
+		else:
+			return X
+
+	def fit(self, X, y, X_test, y_test):
+		"""
+		Takes an input dataset X and a series of targets y and trains the CNN.
+		"""
+		# nodes * channels must = X.shape[1]
+		if self.nodes * self.channels != X.shape[1]:
```
[the remark this answers is not in the document]
Comment (cbarnes7) [reply afqinsight/cnn.py:403]: done

File: afqinsight/cnn.py (hunk 18 above)
Comment (cbarnes7) [reply afqinsight/cnn.py:395]: done

File: afqinsight/cnn.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (richford) [reply afqinsight/cnn.py:365]: I was suggesting doing
```
n_samples, n_features = X.shape
```
instead of
```
n_subjects = X.shape[0]
```
This would unpack the tuple returned by `X.shape` so that we can get `n_samples` and `n_features` all in one line. Then later, we can check that
`n_features == self.n_nodes * self.n_channels`

File: afqinsight/cnn.py (hunk 4 above)
Comment (cbarnes7) [reply afqinsight/cnn.py:464]: I think the weston haven's or cam-can data had a nan in it and was causing super weird errors in testing that took Ariel and I a long time to trace back to a nan. We decided to add this in to address it. We could get rid of this though and instead check for NaNs and raise a cleaner error if there are any depending on what you think.

**Comment by richford** [comment]
Closing in favor of #83
